OpenRA.Mods.Common/Traits/BotModules/McvExpansionManagerBotModule.cs

```diff
@@ -0,0 +1,980 @@
+			 * in order to make expansion worth the time and risk.
+			 */
+			if (firstUndeploy)
+			{
+				/*
+				 * On the first undeploy, we only consider if bot player encounter the early game rush, when bot player
+				 * has less than 2 production buildings (barrack, war factory, etc) or no refinery, MCV won't move and leave everything behind.
+				 */
+				if (world.ActorsHavingTrait<Building>().Count(a => a.Owner == player && Info.ProductionTypes.Contains(a.Info.Name)) >= 2
+					&& world.ActorsHavingTrait<Refinery>().Any(a => a.Owner == player && Info.RefineryTypes.Contains(a.Info.Name)))
+				{
+					var idleconyards = constructionYards.Actors.Where(a => !a.IsDead).ToList();
+
+					if (idleconyards.Count > 0)
+					{
+						bot.QueueOrder(new Order("DeployTransform", idleconyards[0], true));
+						allowfallback = false;
+					}
+				}
+
+				firstUndeploy = false;
+				moveConyardInterval = world.LocalRandom.Next(Info.MoveConyardTick, Info.MoveConyardTick * 2);
+			}
+			else
+			{
+				/*
+				 * On regular undeploy, we only consider that the bot always keep a deployed conyard (so this only works with 2 or more conyards at field),
+				 * and moving MCV should never break the production of refinery.
+				 *
+				 * Because breaking the production of refinery will cause a dead loop of:
+				 * MCV deploys at expansion location -> MCV is producing refinery -> MCV unpacks and interrupts production -> MCV find no refinery here, see this as expansion location and deploy
+				 */
+				var conyards = constructionYards.Actors
+					.Where(a => !a.IsDead).OrderBy(a => a.ActorID)
+					.ToList();
+
+				if (conyards.Count > 1)
+				{
+					var movableMCV = conyards.FirstOrDefault(a => !a.TraitsImplementing<ProductionQueue>()
+					.Any(t => t.Enabled && t.AllQueued().Any(q => Info.RefineryTypes.Contains(q.Item))));
+
+					if (movableMCV != null)
+						bot.QueueOrder(new Order("DeployTransform", movableMCV, true));
+				}
+			}
+		}
+
+		// Find any MCV and deploy them at a sensible location.
+		void DeployMcv(IBot bot, Actor mcv, bool move)
+		{
+			CPos? desiredLocation = null;
+			if (move)
+			{
+				var transformsInfo = mcv.Info.TraitInfo<TransformsInfo>();
+				desiredLocation = ChooseMcvDeployLocation(mcv, transformsInfo.IntoActor, transformsInfo.Offset, allowfallback);
+				if (desiredLocation == null)
+					return;
+
+				bot.QueueOrder(new Order("Move", mcv, Target.FromCell(world, desiredLocation.Value), true));
+
+				allowfallback = true;
+			}
+			else
+				desiredLocation = mcv.Location;
+
+			bot.QueueOrder(new Order("DeployTransform", mcv, true));
+
+			// When we don't have a construction yard, we notify the new location to other traits for defence,
+			// If not, we only notify sometimes, because we are not sure if mcv can successfully deploy at the desired location.
+			// TODO: This could be addressed via INotifyTransform.
+			if ((!constructionYards.Actors.Any(a => !a.IsDead && a.IsInWorld)) || world.LocalRandom.Next(2) > 0)
```

Comment:
```suggestion
			if ((!constructionYards.Actors.Any(a => !a.IsDead)) || world.LocalRandom.Next(2) > 0)
```

this check is redundant as the index already only tracks actors that are in world

Reply:
fixed